### DISCUSSION TITLE
Levels of abstraction for groups of components

### DISCUSSION THREAD

**Discussion by colepoirier** [bevyengine/bevy#4434] [General]
#4072 was migrated here because its content and intent were much more that of a 'discussion' than 'tracking issue'.

This is an issue for tracking the discussions and implementations of the conceptual hierarchy of abstraction for data i.e. `Component`s in Bevy. We need figure out where to draw the lines between the kinds of organization in bevy. The currently existing levels of abstraction are `Component`, `Bundle`, and `Scene`. There is currently discussion of the conceptual clarity and ergonomics of `Bundle`, and `Scene`. There are proposals to change the definition and meaning of `Bundle`, and additional proposals to add new capabilities to it specifically a dynamic/runtime reflection API. There are discussions of the conceptual clarity of `Scene` and proposals to replace it with new and more levels of abstraction called `Schematic`, `Prefab`, and `Blueprint`, or to keep `Scene` while adding new levels of abstraction, narrowing the currently very large scope of `Scene`. These discussion and proposals have wide ranging implications for the architecture of the engine, Cart said that even a seemingly small proposal like the one to implement `Bundle` for all `Components` may have far reaching implications for the engine and has the potential limit or hamstring future development it is not carefully evaluated in a holistic and future-looking manner. Proposals like `Schematics` and dynamic `Components` and dynamic `Bundle` would be primarily useful and perhaps essential for good ergonomics of the Bevy Visual Editor, and they have other potential uses independent of this. 

The most recent and extensive discussion of `Scenes` involving Cart occurred on 28 February 2022 on the bevy discord server in the #scenes-dev channel starting at this message: <https://discord.com/channels/691052431525675048/745805740274614303/948117767754764348>

Issue: Better tools for working with dynamic collections of components #3227

Guidance from @alice-i-cecile for my filling in of this issue: "I think the other core idea that we should communicate is that things should be able to be promoted up the hierarchy of abstraction levels, but not back down."

### What is a `Component`

A `Component` is single struct or enum that implements `Component` which is stored on and `Entity`. 

### What is a `Bundle`

A `Bundle` or properly as Component Bundle is an abstraction that allows you to organize one or more `Component`s in a structs so that a user can spawn a group of `Components` without having to manually insert them on an `Entity` individually. When a user inserts a Component `Bundle`, Bevy turns each of its fields into a distinct component and inserts them on the entity.

The user can nest `Bundles`, embedding one bundle of components within another, because `Bundles` are after all just a collection of `Component`. Bundles are not currently checked (https://github.com/bevyengine/bevy/issues/2387) for duplicate component types. `Component` inserted on an `Entity` will overwrite existing `Components` of the same type.

### What is a `Scene`

Scenes focus area tracking issue #255

### What is proposed to be a `Schematic`

Issue: `Schematic` proposal and discussion #3877

### `DynamicBundle` proposal and discussion

PR: Make a dynamically applicable version of Bundle #3694

### `SceneBundle` proposal and discussion

PR: add a SceneBundle to spawn a scene #2424

### Implement `Bundle` for every `Component` proposal and discussion

PR: Implement Bundle for every Component #2975

### Simplified `Scene` and reflection serialization and proposal

Issue proposing two alternative syntaxes to improve the verbosity and format of `Scene` files #4153

**Comment by SamPruden** [comment]
> Issue: Schematic proposal and discussion https://github.com/bevyengine/bevy/pull/2877

Just FYI that should be #3877.

**Reply by colepoirier** [reply]
Thanks! I guess I didn’t fix it when I copied it over from the original issue.

**Comment by adsick** [comment]
Hi, I'm thinking on universal reusable constructors for "bundles", they have properties of bundles, but the difference is that they also may have children. I wonder how that should be called?

**Reply by adsick** [reply]
I've described it in details in this discussion:  #8375
